File: FilFeed/src/main/java/com/phicdy/filfeed/ui/CurationListFragment.java
package com.phicdy.filfeed.ui;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.phicdy.filfeed.R;
import com.phicdy.filfeed.db.DatabaseAdapter;
import com.phicdy.filfeed.rss.Curation;
import com.phicdy.filfeed.rss.UnreadCountManager;

import java.util.ArrayList;

public class CurationListFragment extends Fragment {

    private CurationListAdapter curationListAdapter;
    private OnCurationListFragmentListener mListener;
    private ListView curationListView;

    private ArrayList<Curation> allCurations = new ArrayList<>();
    private DatabaseAdapter dbAdapter;
    private UnreadCountManager unreadManager;

    private static final String LOG_TAG = "FilFeed.CurationList";

    public static CurationListFragment newInstance() {
        return new CurationListFragment();
    }

    public CurationListFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        unreadManager = UnreadCountManager.getInstance(getActivity());
        dbAdapter = DatabaseAdapter.getInstance(getActivity());
        allCurations = dbAdapter.getAllCurations();
    }

    @Override
    public void onResume() {
        super.onResume();
        refreshList();
    }


    @Override
    public void onPause() {
        super.onPause();
    }

    private void setAllListener() {
        // When an curation selected, display unread articles in the curation
        curationListView
                .setOnItemClickListener(new AdapterView.OnItemClickListener() {

                    @Override
                    public void onItemClick(AdapterView<?> parent, View view,
                                            int position, long id) {
                        mListener.onCurationListClicked(position);
                    }

                });
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_curation_list, container, false);
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        try {
            mListener = (OnCurationListFragmentListener) activity;
        } catch (ClassCastException e) {
            throw new ClassCastException(activity.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        curationListView = (ListView) getActivity().findViewById(R.id.lv_curation);
        TextView emptyView = (TextView)getActivity().findViewById(R.id.emptyView_curation);
        curationListView.setEmptyView(emptyView);
        getActivity().registerForContextMenu(curationListView);
        setAllListener();

        refreshList();

        // Set ListView
        curationListAdapter = new CurationListAdapter(allCurations, getActivity());
        curationListView.setAdapter(curationListAdapter);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public void refreshList() {
        curationListAdapter = new CurationListAdapter(allCurations, getActivity());
        curationListView.setAdapter(curationListAdapter);
        curationListAdapter.notifyDataSetChanged();
    }

    public void addCuration(Curation newCuration) {
        allCurations.add(newCuration);
        curationListAdapter.notifyDataSetChanged();
    }

    public void removeCurationAtPosition(int position) {
        Curation deletedCuration = allCurations.get(position);
        dbAdapter.deleteCuration(deletedCuration.getId());
        allCurations.remove(position);
        curationListAdapter.notifyDataSetChanged();
    }

    public int getCurationIdAtPosition (int position) {
        if (position < 0) {
            return -1;
        }

        if (allCurations == null || position > allCurations.size()-1) {
            return -1;
        }
        return allCurations.get(position).getId();
    }

    public interface OnCurationListFragmentListener {
        public void onCurationListClicked(int position);
    }

    /**
     *
     * @author kyamaguchi Display RSS Curations List
     */
    class CurationListAdapter extends ArrayAdapter<Curation> {
        public CurationListAdapter(ArrayList<Curation> curations, Context context) {
            super(context, R.layout.curation_list, curations);
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            ViewHolder holder;

            // Use contentView and setup ViewHolder
            View row = convertView;
            if (convertView == null) {
                LayoutInflater inflater = getActivity().getLayoutInflater();
                row = inflater.inflate(R.layout.curation_list, parent, false);
                holder = new ViewHolder();
                holder.curationTitle = (TextView) row.findViewById(R.id.tv_curation_title);
                holder.curationCount = (TextView) row.findViewById(R.id.tv_curation_count);
                row.setTag(holder);
            } else {
                holder = (ViewHolder) row.getTag();
            }

            Curation curation = this.getItem(position);

            holder.curationTitle.setText(curation.getTitle());

            // set RSS Curation unread article count
            holder.curationCount.setText(String.valueOf(unreadManager.getUnreadCount(curation.getId())));

            return row;
        }

        private class ViewHolder {
            TextView curationTitle;
            TextView curationCount;
        }
    }

}

File: legacy/src/main/java/com/phicdy/mycuration/presentation/view/fragment/CurationListFragment.kt
package com.phicdy.mycuration.presentation.view.fragment

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.phicdy.mycuration.data.repository.CurationRepository
import com.phicdy.mycuration.data.repository.RssRepository
import com.phicdy.mycuration.entity.Curation
import com.phicdy.mycuration.legacy.R
import com.phicdy.mycuration.presentation.presenter.CurationListPresenter
import com.phicdy.mycuration.presentation.view.CurationItem
import com.phicdy.mycuration.presentation.view.CurationListView
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.components.FragmentComponent
import dagger.hilt.android.scopes.FragmentScoped
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.ArrayList
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

@AndroidEntryPoint
class CurationListFragment : Fragment(), CurationListView, CoroutineScope {

    companion object {
        private const val EDIT_CURATION_MENU_ID = 1
        private const val DELETE_CURATION_MENU_ID = 2
    }

    private lateinit var job: Job
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    @Inject
    lateinit var presenter: CurationListPresenter

    @Inject
    lateinit var rssRepository: RssRepository

    @Inject
    lateinit var curationRepository: CurationRepository

    private lateinit var curationListAdapter: CurationListAdapter
    private var mListener: OnCurationListFragmentListener? = null
    private lateinit var curationRecyclerView: RecyclerView
    private lateinit var emptyView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        job = Job()
    }

    override fun onResume() {
        super.onResume()
        launch {
            presenter.resume()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_curation_list, container, false)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            mListener = context as OnCurationListFragmentListener
        } catch (e: ClassCastException) {
            throw ClassCastException(context.toString() + " must implement OnCurationListFragmentListener")
        }

    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        curationRecyclerView = activity?.findViewById(R.id.rv_curation) as RecyclerView
        emptyView = activity?.findViewById(R.id.emptyView_curation) as TextView
        launch {
            presenter.activityCreated()
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun onDestroy() {
        job.cancel()
        super.onDestroy()
    }

    override fun startEditCurationActivity(editCurationId: Int) {
        mListener?.startEditCurationActivity(editCurationId)
    }

    override fun setNoRssTextToEmptyView() {
        emptyView.setText(R.string.no_rss_message)
    }

    override fun registerContextMenu() {
        registerForContextMenu(curationRecyclerView)
    }

    override fun initListBy(curations: ArrayList<Curation>) {
        curationListAdapter = CurationListAdapter(curations)
        curationRecyclerView.adapter = curationListAdapter
        curationRecyclerView.layoutManager = LinearLayoutManager(activity)
        curationListAdapter.notifyDataSetChanged()
    }

    override fun showRecyclerView() {
        curationRecyclerView.visibility = View.VISIBLE
    }

    override fun hideRecyclerView() {
        curationRecyclerView.visibility = View.GONE
    }

    override fun showEmptyView() {
        emptyView.visibility = View.VISIBLE
    }

    override fun hideEmptyView() {
        emptyView.visibility = View.GONE
    }

    interface OnCurationListFragmentListener {
        fun onCurationListClicked(curationId: Int)
        fun startEditCurationActivity(editCurationId: Int)
    }

    inner class CurationListAdapter(private val curations: ArrayList<Curation>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            val itemView = LayoutInflater.from(parent.context).inflate(R.layout.curation_list, parent, false)
            return ViewHolder(itemView)
        }

        override fun getItemCount(): Int {
            return curations.size
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            holder.itemView.setOnClickListener {
                val curation = curations[position]
                mListener?.onCurationListClicked(curation.id)
            }

            holder.itemView.setOnCreateContextMenuListener { menu, _, _ ->
                menu.add(0, EDIT_CURATION_MENU_ID, 0, R.string.edit_curation).setOnMenuItemClickListener {
                    val editCuration = curations[position]
                    presenter.onCurationEditClicked(editCuration.id)
                    true
                }
                menu.add(0, DELETE_CURATION_MENU_ID, 0, R.string.delete_curation).setOnMenuItemClickListener {
                    launch {
                        presenter.onCurationDeleteClicked(curations[position], curations.size)
                        curations.removeAt(position)
                        notifyItemRemoved(position)
                        delay(500) // Wait for delete animation
                        notifyDataSetChanged()
                    }
                    true
                }
            }
            val curation = curations[position]
            launch {
                presenter.getView(curation, holder as CurationItem)
            }

        }

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), CurationItem {
            val curationName = itemView.findViewById(R.id.tv_curation_title) as TextView
            val curationCount = itemView.findViewById(R.id.tv_curation_count) as TextView

            override fun setName(name: String) {
                curationName.text = name
            }

            override fun setCount(count: String) {
                curationCount.text = count
            }
        }
    }

    @Module
    @InstallIn(FragmentComponent::class)
    object CurationListModule {
        @FragmentScoped
        @Provides
        fun provideCurationListView(fragment: Fragment): CurationListView = fragment as CurationListView
    }
}
